biz.aQute.bndlib/src/aQute/bnd/cdi/CDIAnnotationReader.java

```diff
@@ -0,0 +1,542 @@
+			ReferenceTypeSignature inferred = resolver.resolveType(type.classType.typeArguments[0]);
+			if (!(inferred instanceof ClassTypeSignature)) {
+				analyzer.error(
+					"In bean %s, in member %s with @Reference the type argument of Provider can't be resolved: %s",
+					clazz,
+					member.getName(), inferred);
+				return;
+			}
+			type = (ClassTypeSignature) inferred;
+			fqn = Descriptors.binaryToFQN(type.binary);
+		}
+
+		// unwrap Collection, List, Optional
+		if (fqn.equals("java.util.Collection") || fqn.equals("java.util.List")) {
+			ReferenceTypeSignature inferred = resolver.resolveType(type.classType.typeArguments[0]);
+			if (!(inferred instanceof ClassTypeSignature)) {
+				analyzer.error(
+					"In bean %s, in member %s with @Reference the type argument of Collection or List can't be resolved: %s",
+					clazz, member.getName(), inferred);
+				return;
+			}
+			type = (ClassTypeSignature) inferred;
+			fqn = Descriptors.binaryToFQN(type.binary);
+			cardinality = ReferenceCardinality.MULTIPLE;
+		} else if (fqn.equals("java.util.Optional")) {
+			ReferenceTypeSignature inferred = resolver.resolveType(type.classType.typeArguments[0]);
+			if (!(inferred instanceof ClassTypeSignature)) {
+				analyzer.error(
+					"In bean %s, in member %s with @Reference the type argument of Optional can't be resolved: %s",
+					clazz, member.getName(), inferred);
+				return;
+			}
+			type = (ClassTypeSignature) inferred;
+			fqn = Descriptors.binaryToFQN(type.binary);
+			cardinality = ReferenceCardinality.OPTIONAL;
+		}
+
+		if (fqn.equals("org.osgi.service.cdi.reference.BeanServiceObjects")
+			|| fqn.equals("org.osgi.framework.ServiceReference")) {
+
+			ReferenceTypeSignature inferred = resolver.resolveType(type.classType.typeArguments[0]);
+			if (!(inferred instanceof ClassTypeSignature)) {
+				analyzer.error(
+					"In bean %s, in member %s with @Reference the type argument of BeanServiceObjects or ServiceReference can't be resolved: %s",
+					clazz, member.getName(), inferred);
+				return;
+			}
+			type = (ClassTypeSignature) inferred;
+			fqn = Descriptors.binaryToFQN(type.binary);
+		} else if (fqn.equals("java.util.Map$Entry")) {
+			ReferenceTypeSignature inferred = resolver.resolveType(type.innerTypes[0].typeArguments[1]);
+			if (!(inferred instanceof ClassTypeSignature)) {
+				analyzer.error(
+					"In bean %s, in member %s with @Reference the second type argument of Map.Entry can't be resolved: %s",
+					clazz, member.getName(), inferred);
+				return;
+			}
+			type = (ClassTypeSignature) inferred;
+			fqn = Descriptors.binaryToFQN(type.binary);
+		}
+
+		referenceDef = new ReferenceDef();
+		referenceDef.service = fqn;
+		referenceDef.cardinality = cardinality;
+		definitions.get(0).references.add(referenceDef);
+	}
+
+	private void doService(Annotation annotation, java.lang.annotation.Annotation a) {
+		switch (annotation.getElementType()) {
+			case FIELD :
+				if (member instanceof Clazz.FieldDef) {
```

Comment:
I think this instanceof test in unnecessary. I have a change to the DS processing to rely strictly on annotation.getElementType().